netebpfext/net_ebpf_ext.c

```diff
@@ -28,6 +28,11 @@ NDIS_HANDLE _net_ebpf_ext_ndis_handle = NULL;
 NDIS_HANDLE _net_ebpf_ext_nbl_pool_handle = NULL;
 HANDLE _net_ebpf_ext_l2_injection_handle = NULL;
 
+static bool _net_ebpf_xdp_providers_registered = false;
+static bool _net_ebpf_bind_providers_registered = false;
+static bool _net_ebpf_sock_addr_providers_registered = false;
+static bool _net_ebpf_sock_ops_providers_registered = false;
```

Comment:
Can all these variables be defined in their respective files:

For example,
In file,
  net_ebpf_ext_xdp.c

static bool _net_ebpf_xdp_set = false;

NTSTATUS
net_ebpf_ext_xdp_set_providers_registered() or net_ebpf_ext_xdp_register_providers()
{
   _net_ebpf_xdp_providers_registered = true;
  
}
void
net_ebpf_ext_xdp_unregister_providers()
{
    net_ebpf_extension_hook_provider_unregister(_ebpf_xdp_hook_provider_context);
    net_ebpf_extension_program_info_provider_unregister(_ebpf_xdp_program_info_provider_context);
   _net_ebpf_xdp_providers_registered = false;
}

Reply:
A better design would not need these bools at all, but would instead NULL out `_ebpf_xdp_hook_provider_context` etc.
Ideally, such NULLing would be done inside net_ebpf_extension_hook_provider_unregister(&_ebpf_xdp_hook_provider_context) and similar unregister functions, which would make this one step closer to RAII.

Reply:
The `unregister` functions are not idempotent, particularly `net_ebpf_ext_sock_addr_unregister_providers()`. So there needs to be a check before `net_ebpf_ext_sock_addr_unregister_providers()` does the cleanup (either inside this function or in the caller). We cannot have the check for the "initialized" bool value inside `net_ebpf_ext_sock_addr_unregister_providers()` as we want to be able to call `net_ebpf_ext_sock_addr_unregister_providers()` in the failure path of `net_ebpf_ext_sock_addr_register_providers()` also, and the bool value will not be true in that case.

Reply:
@dthaler I had actually tried removing the bools in this commit
https://github.com/microsoft/ebpf-for-windows/pull/2170/commits/aef03733b6a56e5da91ab76fdc82d2de7c066e7f

But the tests failed -- `net_ebpf_ext_sock_addr_unregister_providers()` not only unregisters the provider context (which I am now tracking by checking and setting them to NULL), but also purges entries from a list. If the list is not initialized in the first place (which happened in the low memory test -- `net_ebpf_ext_register_providers()` bailed as `net_ebpf_ext_bind_register_providers()` failed, so `net_ebpf_ext_sock_addr_register_providers()` was never called, hence the linked list was never initialized, but then `net_ebpf_ext_unregister_providers()` unconditionally called `net_ebpf_ext_sock_addr_unregister_providers()`), and then `net_ebpf_ext_sock_addr_unregister_providers()` is called, it crashes while traversing the list.